pandas 2.0.0 brings some warnings and incompatible changes

pandas 2.0.0 was released on Apr 3, 2023 and it brings some incompatible changes (see https://pandas.pydata.org/docs/whatsnew/v2.0.0.html for the changelog). 

Here are the new failures and warnings with pandas 2.0.0 (https://github.com/GenericMappingTools/pygmt/actions/runs/4610492623/jobs/8149020030?pr=2477):

```
=================================== FAILURES ===================================
______________ [doctest] pygmt.clib.conversion.array_to_datetime _______________
276 
277     Examples
278     --------
279     >>> import datetime
280     >>> # numpy.datetime64 array
281     >>> x = np.array(
282     ...     ["2010-06-01", "2011-06-01T12", "2012-01-01T12:34:56"],
283     ...     dtype="datetime64",
284     ... )
285     >>> array_to_datetime(x)
Differences (unified diff with -expected +actual):
    @@ -1,3 +1,3 @@
     DatetimeIndex(['2010-06-01 00:00:00', '2011-06-01 12:00:00',
                    '2012-01-01 12:34:56'],
    -              dtype='datetime64[ns]', freq=None)
    +              dtype='datetime64[s]', freq=None)

/home/runner/work/pygmt/pygmt/pygmt/clib/conversion.py:285: DocTestFailure
=============================== warnings summary ===============================
pygmt/helpers/decorators.py::pygmt.helpers.decorators.kwargs_to_strings
  <doctest pygmt.helpers.decorators.kwargs_to_strings[13]>:3: UserWarning: Converting non-nanosecond precision datetime values to nanosecond precision. This behavior can eventually be relaxed in xarray, as it is an artifact from pandas which is now beginning to support non-nanosecond precision values. This warning is caused by passing non-nanosecond np.datetime64 or np.timedelta64 values to the DataArray or Variable constructor; it can be silenced by converting the values to nanosecond precision ahead of time.

pygmt/tests/test_x2sys_cross.py: 12 warnings
  /home/runner/work/pygmt/pygmt/pygmt/src/x2sys_cross.py:233: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
    table = pd.read_csv(
```

--- a/.github/workflows/ci_tests_legacy.yaml
+++ b/.github/workflows/ci_tests_legacy.yaml
@@ -11,7 +11,7 @@ name: GMT Legacy Tests
 on:
   # push:
   #   branches: [ main ]
-  pull_request:
+  # pull_request:
     # types: [ready_for_review]
     # paths-ignore:
     #  - 'doc/**'

--- a/.github/workflows/ci_tests_legacy.yaml
+++ b/.github/workflows/ci_tests_legacy.yaml
@@ -11,7 +11,7 @@ name: GMT Legacy Tests
 on:
   # push:
   #   branches: [ main ]
-  # pull_request:
+  pull_request:
     # types: [ready_for_review]
     # paths-ignore:
     #  - 'doc/**'
@@ -67,7 +67,7 @@ jobs:
             python=3.9
             gmt=${{ matrix.gmt_version }}
             numpy
-            pandas
+            pandas<2
             xarray
             netCDF4
             packaging

--- a/pygmt/src/x2sys_cross.py
+++ b/pygmt/src/x2sys_cross.py
@@ -241,7 +241,7 @@ def x2sys_cross(tracks=None, outfile=None, **kwargs):
                     header=2,  # Column names are on 2nd row
                     comment=">",  # Skip the 3rd row with a ">"
                     parse_dates=[2, 3],  # Datetimes on 3rd and 4th column
-                    *date_format_kwarg,  # Parse dates in ISO8601 format on pandas>=2
+                    **date_format_kwarg,  # Parse dates in ISO8601 format on pandas>=2
                 )
                 # Remove the "# " from "# x" in the first column
                 table = table.rename(columns={table.columns[0]: table.columns[0][2:]})

--- a/pygmt/src/x2sys_cross.py
+++ b/pygmt/src/x2sys_cross.py
@@ -6,6 +6,7 @@ import os
 from pathlib import Path
 
 import pandas as pd
+from packaging.version import Version
 from pygmt.clib import Session
 from pygmt.exceptions import GMTInvalidInput
 from pygmt.helpers import (
@@ -229,13 +230,18 @@ def x2sys_cross(tracks=None, outfile=None, **kwargs):
             # Read temporary csv output to a pandas table
             if outfile == tmpfile.name:  # if outfile isn't set, return pd.DataFrame
                 # Read the tab-separated ASCII table
+                date_format_kwarg = (
+                    {"date_format": "ISO8601"}
+                    if Version(pd.__version__) >= Version("2.0.0")
+                    else {}
+                )
                 table = pd.read_csv(
                     tmpfile.name,
                     sep="\t",
                     header=2,  # Column names are on 2nd row
                     comment=">",  # Skip the 3rd row with a ">"
                     parse_dates=[2, 3],  # Datetimes on 3rd and 4th column
-                    date_format="ISO8601",
+                    *date_format_kwarg,  # Parse dates in ISO8601 format on pandas>=2
                 )
                 # Remove the "# " from "# x" in the first column
                 table = table.rename(columns={table.columns[0]: table.columns[0][2:]})

--- a/pygmt/src/x2sys_cross.py
+++ b/pygmt/src/x2sys_cross.py
@@ -47,7 +47,7 @@ def tempfile_from_dftrack(track, suffix):
             sep="\t",
             index=False,
             na_rep="NaN",  # write a NaN value explicitly instead of a blank string
-            date_format="%Y-%m-%dT%H:%M:%S.%fZ",
+            date_format="%Y-%m-%dT%H:%M:%S.%fZ",  # ISO8601 format
         )
         yield tmpfilename
     finally:
@@ -235,6 +235,7 @@ def x2sys_cross(tracks=None, outfile=None, **kwargs):
                     header=2,  # Column names are on 2nd row
                     comment=">",  # Skip the 3rd row with a ">"
                     parse_dates=[2, 3],  # Datetimes on 3rd and 4th column
+                    date_format="ISO8601",
                 )
                 # Remove the "# " from "# x" in the first column
                 table = table.rename(columns={table.columns[0]: table.columns[0][2:]})

--- a/pygmt/tests/test_x2sys_cross.py
+++ b/pygmt/tests/test_x2sys_cross.py
@@ -112,7 +112,7 @@ def test_x2sys_cross_input_two_dataframes(mock_x2sys_home):
         for i in range(2):
             np.random.seed(seed=i)
             track = pd.DataFrame(data=np.random.rand(10, 3), columns=("x", "y", "z"))
-            track["time"] = pd.date_range(start=f"2020-{i}1-01", periods=10, freq="ms")
+            track["time"] = pd.date_range(start=f"2020-{i}1-01", periods=10, freq="min")
             tracks.append(track)
 
         output = x2sys_cross(tracks=tracks, tag=tag, coe="e")

